src/dmd/expressionsem.d

```diff
@@ -4333,6 +4333,9 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             printf("AssertExp::semantic('%s')\n", exp.toChars());
         }
 
+        // save expression as a a string before any semantic expansion
+        auto assertExpMsg = exp.msg ? null : exp.toChars();
```

Comment:
Right, this should be done before semantic, but is using `toChars` correct here? What happens with multiline expressions?

Reply:
Could `exp.copy()` be used instead?

Reply:
isn't `exp.copy()` shallow, so that the copy will still be affected by semantic analysis?